File: smoke-test.py
#!/usr/bin/env python3
"""
HQ Deploy Smoke Test
Runs after every deploy to verify all sites and data integrity.

Usage:
  python smoke-test.py          # Full test (network + local)
  python smoke-test.py --local  # Local data.json validation only (no network)
"""
import json, sys, os, time

RED = '\033[91m'
GREEN = '\033[92m'
GOLD = '\033[93m'
RESET = '\033[0m'
CHECK = f'{GREEN}PASS{RESET}'
CROSS = f'{RED}FAIL{RESET}'

results = []
LOCAL_ONLY = '--local' in sys.argv

def test(name, fn):
    try:
        ok, detail = fn()
        status = CHECK if ok else CROSS
        results.append(ok)
        print(f'  [{status}] {name}: {detail}')
    except Exception as e:
        results.append(False)
        print(f'  [{CROSS}] {name}: {RED}{e}{RESET}')

# ===== LOCAL VALIDATION =====
def validate_local_json():
    """Validate local data.json — no network needed"""
    script_dir = os.path.dirname(os.path.abspath(__file__))
    path = os.path.join(script_dir, 'data.json')

    if not os.path.exists(path):
        return False, f'File not found: {path}'

    raw = open(path, 'rb').read()

    # Null byte check
    nulls = [i for i, b in enumerate(raw) if b == 0]
    if nulls:
        return False, f'NULL BYTES at positions: {nulls}'

    # JSON parse
    try:
        data = json.loads(raw)
    except json.JSONDecodeError as e:
        return False, f'JSON parse error: {e}'

    # Required keys
    required = ['stats', 'roadmap', 'deploys', 'portfolio', 'quickActions', 'tools', 'aiTools', 'hoursLog', 'systems', 'ventures', 'lastUpdated']
    missing = [k for k in required if k not in data]
    if missing:
        return False, f'Missing keys: {missing}'

    # Stats type check
    for field in ['liveSites', 'tasksDone', 'deploys']:
        val = data.get('stats', {}).get(field)
        if not isinstance(val, (int, float)):
            return False, f'stats.{field} invalid: {val}'

    # Non-empty arrays
    arrays = {k: len(data[k]) for k in ['portfolio', 'quickActions', 'tools', 'systems', 'deploys', 'ventures']}
    empty = [k for k, v in arrays.items() if v == 0]
    if empty:
        return False, f'Empty arrays: {empty}'

    # Portfolio URLs present
    for i, p in enumerate(data['portfolio']):
        for key in ['name', 'url', 'emoji', 'desc', 'tags']:
            if key not in p:
                return False, f'portfolio[{i}] missing "{key}"'

    # Roadmap structure
    for phase in ['p1', 'p2', 'p3', 'p4']:
        if phase not in data['roadmap']:
            return False, f'roadmap.{phase} missing'
        if 'progress' not in data['roadmap'][phase]:
            return False, f'roadmap.{phase}.progress missing'

    summary = f'{arrays["portfolio"]} sites, {data["stats"]["tasksDone"]} tasks, {arrays["deploys"]} deploys, updated {data["lastUpdated"]}'
    return True, summary

def check_file_size():
    """Make sure data.json isn't suspiciously small or huge"""
    path = os.path.join(os.path.dirname(os.path.abspath(__file__)), 'data.json')
    size = os.path.getsize(path)
    if size < 500:
        return False, f'{size} bytes — suspiciously small, possible truncation'
    if size > 50000:
        return False, f'{size} bytes — unusually large, check for bloat'
    return True, f'{size} bytes'

def check_index_html():
    """Validate index.html basics"""
    path = os.path.join(os.path.dirname(os.path.abspath(__file__)), 'index.html')
    if not os.path.exists(path):
        return False, 'File not found'

    raw = open(path, 'rb').read()
    if b'\x00' in raw:
        return False, 'NULL BYTES detected'

    content = raw.decode('utf-8', errors='replace')
    checks = {
        'DOCTYPE': '<!DOCTYPE html>' in content,
        'GSAP CDN': 'gsap' in content.lower(),
        'data.json fetch': "fetch('data.json" in content or 'fetch("data.json' in content,
        'error fallback': 'Data sync failed' in content,
        'counter system': 'animateCounter' in content,
        'visibility handler': 'visibilitychange' in content,
    }
    failed = [k for k, v in checks.items() if not v]
    if failed:
        return False, f'Missing: {", ".join(failed)}'
    return True, f'All {len(checks)} checks pass ({len(content)} chars)'

# ===== NETWORK TESTS =====
def check_url(url, timeout=10):
    from urllib.request import urlopen, Request
    req = Request(url, headers={'User-Agent': 'HQ-SmokeTest/1.0'})
    r = urlopen(req, timeout=timeout)
    return r.status == 200, f'{r.status} OK'

def check_remote_json():
    from urllib.request import urlopen, Request
    req = Request(f'https://sean-hq.vercel.app/data.json?t={int(time.time())}',
                  headers={'User-Agent': 'HQ-SmokeTest/1.0'})
    raw = urlopen(req, timeout=10).read()
    if b'\x00' in raw:
        return False, 'NULL BYTES in remote data.json'
    data = json.loads(raw)
    return True, f'Remote valid — {len(data["portfolio"])} sites, {data["stats"]["tasksDone"]} tasks'

def check_portfolio_live():
    from urllib.request import urlopen, Request
    req = Request(f'https://sean-hq.vercel.app/data.json?t={int(time.time())}',
                  headers={'User-Agent': 'HQ-SmokeTest/1.0'})
    data = json.loads(urlopen(req, timeout=10).read())
    statuses = []
    all_ok = True
    for site in data['portfolio']:
        try:
            sreq = Request(site['url'], headers={'User-Agent': 'HQ-SmokeTest/1.0'})
            r = urlopen(sreq, timeout=10)
            statuses.append(f'{site["name"]}=OK')
        except Exception:
            statuses.append(f'{site["name"]}=DOWN')
            all_ok = False
    return all_ok, ' | '.join(statuses)

# ===== RUN =====
print(f'\n{GOLD}{"━" * 40}')
print(f'  HQ SMOKE TEST {"(LOCAL)" if LOCAL_ONLY else "(FULL)"}')
print(f'{"━" * 40}{RESET}\n')

print(f'{GOLD}[Local Validation]{RESET}')
test('data.json integrity', validate_local_json)
test('data.json file size', check_file_size)
test('index.html integrity', check_index_html)

if not LOCAL_ONLY:
    print(f'\n{GOLD}[Network — Live Sites]{RESET}')
    test('HQ Dashboard', lambda: check_url('https://sean-hq.vercel.app/'))
    test('Trinh Media', lambda: check_url('https://trinh-media-site.vercel.app/'))
    test('T4 Folsom', lambda: check_url('https://t4-folsom.vercel.app/'))

    print(f'\n{GOLD}[Network — Data]{RESET}')
    test('Remote data.json', check_remote_json)
    test('Portfolio sites live', check_portfolio_live)

# ===== SUMMARY =====
passed = sum(results)
total = len(results)
print(f'\n{GOLD}{"━" * 40}{RESET}')
if all(results):
    print(f'  {GREEN}ALL PASS ({passed}/{total}){RESET}')
else:
    print(f'  {RED}{total - passed} FAILED / {total} total{RESET}')
print(f'{GOLD}{"━" * 40}{RESET}\n')

sys.exit(0 if all(results) else 1)

File: smoke-test.js
#!/usr/bin/env node
/**
 * HQ Deploy Smoke Test
 * Runs after every deploy to verify all sites and data integrity.
 *
 * Usage:
 *   node smoke-test.js          # Full test (network + local)
 *   node smoke-test.js --local  # Local data.json validation only
 */
const fs = require('fs');
const path = require('path');
const https = require('https');

const RED = '\x1b[91m', GREEN = '\x1b[92m', GOLD = '\x1b[93m', RESET = '\x1b[0m';
const PASS = `${GREEN}PASS${RESET}`, FAIL = `${RED}FAIL${RESET}`;
const LOCAL_ONLY = process.argv.includes('--local');
const results = [];

function test(name, fn) {
    return fn().then(([ok, detail]) => {
        results.push(ok);
        console.log(`  [${ok ? PASS : FAIL}] ${name}: ${detail}`);
    }).catch(e => {
        results.push(false);
        console.log(`  [${FAIL}] ${name}: ${RED}${e.message}${RESET}`);
    });
}

function fetchURL(url, timeout = 10000) {
    return new Promise((resolve, reject) => {
        const req = https.get(url, { headers: { 'User-Agent': 'HQ-SmokeTest/1.0' }, timeout }, res => {
            const chunks = [];
            res.on('data', c => chunks.push(c));
            res.on('end', () => resolve({ status: res.statusCode, body: Buffer.concat(chunks) }));
        });
        req.on('error', reject);
        req.on('timeout', () => { req.destroy(); reject(new Error('Timeout')); });
    });
}

// ===== LOCAL TESTS =====
async function validateLocalJSON() {
    const filePath = path.join(__dirname, 'data.json');
    if (!fs.existsSync(filePath)) return [false, 'File not found'];

    const raw = fs.readFileSync(filePath);

    // Null byte check
    const nullPositions = [];
    for (let i = 0; i < raw.length; i++) { if (raw[i] === 0) nullPositions.push(i); }
    if (nullPositions.length) return [false, `NULL BYTES at positions: ${nullPositions}`];

    let data;
    try { data = JSON.parse(raw); }
    catch (e) { return [false, `JSON parse error: ${e.message}`]; }

    // Required keys
    const required = ['stats','roadmap','deploys','portfolio','quickActions','tools','aiTools','systems','ventures','lastUpdated'];
    const missing = required.filter(k => !(k in data));
    if (missing.length) return [false, `Missing keys: ${missing}`];

    // Stats type check
    for (const f of ['liveSites','tasksDone','deploys']) {
        const val = data.stats?.[f];
        if (typeof val !== 'number') return [false, `stats.${f} invalid: ${val}`];
    }

    // Non-empty arrays
    const arrays = { portfolio: data.portfolio, quickActions: data.quickActions, tools: data.tools, systems: data.systems, deploys: data.deploys, ventures: data.ventures };
    const empty = Object.entries(arrays).filter(([,v]) => !v?.length).map(([k]) => k);
    if (empty.length) return [false, `Empty arrays: ${empty}`];

    // Portfolio field check
    for (let i = 0; i < data.portfolio.length; i++) {
        for (const key of ['name','url','emoji','desc','tags']) {
            if (!(key in data.portfolio[i])) return [false, `portfolio[${i}] missing "${key}"`];
        }
    }

    // Roadmap structure
    for (const p of ['p1','p2','p3','p4']) {
        if (!data.roadmap?.[p]?.progress) return [false, `roadmap.${p}.progress missing`];
    }

    return [true, `Valid — ${data.portfolio.length} sites, ${data.stats.tasksDone} tasks, ${data.deploys.length} deploys`];
}

async function checkFileSize() {
    const filePath = path.join(__dirname, 'data.json');
    const size = fs.statSync(filePath).size;
    if (size < 500) return [false, `${size} bytes — suspiciously small`];
    if (size > 50000) return [false, `${size} bytes — unusually large`];
    return [true, `${size} bytes`];
}

async function checkIndexHTML() {
    const filePath = path.join(__dirname, 'index.html');
    if (!fs.existsSync(filePath)) return [false, 'File not found'];
    const raw = fs.readFileSync(filePath);
    for (let i = 0; i < raw.length; i++) { if (raw[i] === 0) return [false, 'NULL BYTES detected']; }
    const content = raw.toString('utf-8');
    const checks = {
        'DOCTYPE': content.includes('<!DOCTYPE html>'),
        'GSAP CDN': content.toLowerCase().includes('gsap'),
        'data.json fetch': content.includes("fetch('data.json") || content.includes('fetch("data.json'),
        'error fallback': content.includes('Data sync failed'),
        'counter system': content.includes('animateCounter'),
        'visibility handler': content.includes('visibilitychange'),
    };
    const failed = Object.entries(checks).filter(([,v]) => !v).map(([k]) => k);
    if (failed.length) return [false, `Missing: ${failed.join(', ')}`];
    return [true, `All ${Object.keys(checks).length} checks pass (${content.length} chars)`];
}

// ===== NETWORK TESTS =====
async function checkURL(url) {
    const res = await fetchURL(url);
    return [res.status === 200, `${res.status} OK`];
}

async function checkRemoteJSON() {
    const res = await fetchURL(`https://sean-hq.vercel.app/data.json?t=${Date.now()}`);
    for (let i = 0; i < res.body.length; i++) { if (res.body[i] === 0) return [false, 'NULL BYTES in remote']; }
    const data = JSON.parse(res.body);
    return [true, `Remote valid — ${data.portfolio.length} sites, ${data.stats.tasksDone} tasks`];
}

async function checkPortfolioLive() {
    const res = await fetchURL(`https://sean-hq.vercel.app/data.json?t=${Date.now()}`);
    const data = JSON.parse(res.body);
    const statuses = [];
    let allOk = true;
    for (const site of data.portfolio) {
        try {
            const r = await fetchURL(site.url);
            statuses.push(`${site.name}=${r.status === 200 ? 'OK' : r.status}`);
            if (r.status !== 200) allOk = false;
        } catch {
            statuses.push(`${site.name}=DOWN`);
            allOk = false;
        }
    }
    return [allOk, statuses.join(' | ')];
}

// ===== RUN =====
async function main() {
    const bar = '━'.repeat(40);
    console.log(`\n${GOLD}${bar}`);
    console.log(`  HQ SMOKE TEST ${LOCAL_ONLY ? '(LOCAL)' : '(FULL)'}`);
    console.log(`${bar}${RESET}\n`);

    console.log(`${GOLD}[Local Validation]${RESET}`);
    await test('data.json integrity', validateLocalJSON);
    await test('data.json file size', checkFileSize);
    await test('index.html integrity', checkIndexHTML);

    if (!LOCAL_ONLY) {
        console.log(`\n${GOLD}[Network — Live Sites]${RESET}`);
        await test('HQ Dashboard', () => checkURL('https://sean-hq.vercel.app/'));
        await test('Trinh Media', () => checkURL('https://trinh-media-site.vercel.app/'));
        await test('T4 Folsom', () => checkURL('https://t4-folsom.vercel.app/'));

        console.log(`\n${GOLD}[Network — Data]${RESET}`);
        await test('Remote data.json', checkRemoteJSON);
        await test('Portfolio sites live', checkPortfolioLive);
    }

    const passed = results.filter(Boolean).length;
    console.log(`\n${GOLD}${bar}${RESET}`);
    if (results.every(Boolean)) {
        console.log(`  ${GREEN}ALL PASS (${passed}/${results.length})${RESET}`);
    } else {
        console.log(`  ${RED}${results.length - passed} FAILED / ${results.length} total${RESET}`);
    }
    console.log(`${GOLD}${bar}${RESET}\n`);

    process.exit(results.every(Boolean) ? 0 : 1);
}

main();
